### PR TITLE
Overhaul tools and tutorials

### PR DESCRIPTION
**Pull request by gudbrandtandberg**

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,77 @@
+[tool.poetry]
+name = "3lc_tools"
+version = "0.1.0"
```
[the remark this answers is not in the document]
Comment (fredmell) [reply pyproject.toml:3]: Most likely the poetry default version - conscious decision making is probably in order.

File: tutorials/3-training-and-metrics/sam-embeddings.ipynb [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "\u001b[90m3lc: \u001b[0mCreated new run at C:/Users/gudbrand/AppData/Local/3LC/3LC/projects/3LC Tutorials/runs/Collect embeddings\n"
```
[the remark this answers is not in the document]
Comment (fredmell) [reply tutorials/3-training-and-metrics/sam-embeddings.ipynb:190]: We should probably run nbformat or a similar tool to do this

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
+# Contributing to 3lc-examples
+
+Thank you for your interest in contributing to our collection of tutorial notebooks! We welcome contributions from the community to enhance and expand our repository. Below are the guidelines to help you contribute effectively.
```
[the remark this answers is not in the document]
Comment (gudbrandtandberg) [reply CONTRIBUTING.md:3]: Good point. I think we should not rule out contributions to tools!

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+description = "A set of open-source tools for extending and enhancing the base 3lc functionality."
+authors = ["3LC <hello@3lc.ai>"]
```
[the remark this answers is not in the document]
Comment (fredmell) [reply pyproject.toml]: I don't think it is a problem to have it there - updated to the real one.

File: src/tlc_tools/embeddings.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    embedding_column: str = "embedding",
+    batch_size: int = 4,
+    device: str | torch.device | None = None,
+    preprocess_fn: Callable | None = None,
+) -> tlc.Table:
+    """
+    Adds embeddings to a table using a specified model and optional dimensionality reduction.
+
+
+    :param table: The table containing images.
```
[the remark this answers is not in the document]
Comment (fredmell) [reply src/tlc_tools/embeddings.py:31]: Went with :param ...: and :returns:

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  ```html
+    <div style="background-color: #e7f3fe; padding: 10px; border-left: 6px solid #2196F3; margin-bottom: 15px; color: #333;">,
+      <strong>Info:</strong> This is an info box.
+      <ul>
+          <li>There are several pieces of information here.</li>
+          <li>Here is another one</li>
+      </ul>
+    </div>
+  ```
+
```
[the remark this answers is not in the document]
Comment (gudbrandtandberg) [reply CONTRIBUTING.md:74]: Added screenshots and more examples.

File: CONTRIBUTING.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (gudbrandtandberg) [reply CONTRIBUTING.md:3]: Changed wording.

File: src/tlc_tools/split.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+
+class _TraversalIndexSplitStrategy(_RandomSplitStrategy):
+    requires_split_by = True
+
+    def split(
+        self, indices: np.array, splits: dict[str, float], by_column: np.array | None = None
+    ) -> dict[str, np.array]:
+        # Sort to take smallest splits first
+        splits = dict(sorted(splits.items(), key=lambda x: x[1]))
```
[the remark this answers is not in the document]
Comment (fredmell) [reply src/tlc_tools/split.py:71]: Done in split_table!

File: pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+readme = "README.md"
+license = "Apache-2.0"
+keywords = ["AI", "tools", "data workflows", "3lc"]
+packages = [{ include = "tlc_tools", from = "src" }]
+
+[tool.poetry.dependencies]
+python = ">=3.9,<3.13"
```
[the remark this answers is not in the document]
Comment (gudbrandtandberg) [reply pyproject.toml:12]: We do not want to support 3.8. It would be possible, but non-trivial, we really strongly want to drop 3.8 and it is just a matter of time.

File: pyproject.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (gudbrandtandberg) [reply pyproject.toml]: We have had some talks, and we think it makes sense to tie the version to the tlc versions. Although there are some ramifications that need to be discussed here: how to accept contributing PRs, do we need to retain backwards compat etc. We need to have some discussions, lets do it after we release the first overhaul. There are also some more fundamental general questions about what this tlc_tools repo really _is_ in the first place, and what it should be, if anything..

File: src/tlc_tools/add_columns_to_table.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+from collections import defaultdict
+from typing import Any, Callable
+
+import tlc
+from PIL import Image
+from tqdm.auto import tqdm
+
+from tlc_tools.metrics import IMAGE_METRICS, compute_image_metrics
+
+_SampleTypeStructure = typing.Union[
```
[the remark this answers is not in the document]
Comment (gudbrandtandberg) [reply src/tlc_tools/add_columns_to_table.py:13]: For this specific complex type we need to use the Union syntax. In general we prefer |

File: src/tlc_tools/datasets.py [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+        is_background = self.add_background and self.random_gen.random() < self.background_freq and self.is_train
+
+        # Select a random row for background or use the given index
+        row_idx = self.random_gen.randint(0, len(self.table) - 1) if is_background else idx
+
+        row = self.table.table_rows[row_idx]
+        image = self.load_image_data(row)
+
+        bbs = row["bbs"]["bb_list"]
+        while len(bbs) == 0 and not is_background:
```
[the remark this answers is not in the document]
Comment (gudbrandtandberg) [reply src/tlc_tools/datasets.py]: Yes, leaving the assumption that many/most rows have bbs for now. Note taken.

File: src/tlc_tools/datasets.py [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+        else:
+            crop, label = self.generate_bb_crop(image, bbs)
+
+        if self.transform:
+            crop = self.transform(crop)
+
+        return crop, label
+
+    def load_image_data(self, row):
+        image_bytes = tlc.Url(row["image"]).read()
```
[the remark this answers is not in the document]
Comment (gudbrandtandberg) [reply src/tlc_tools/datasets.py:90]: Made it more flexible!

File: src/tlc_tools/datasets.py [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+        # Determine if a background patch should be generated
+        is_background = self.add_background and self.random_gen.random() < self.background_freq and self.is_train
+
+        # Select a random row for background or use the given index
+        row_idx = self.random_gen.randint(0, len(self.table) - 1) if is_background else idx
+
+        row = self.table.table_rows[row_idx]
+        image = self.load_image_data(row)
+
+        bbs = row["bbs"]["bb_list"]
```
[the remark this answers is not in the document]
Comment (gudbrandtandberg) [reply src/tlc_tools/datasets.py:72]: Yeah, agree with your point, I'll put this in the same to-do-later item as adding more stringent schema-checking and flexible naming stuff.

[the remark this answers is not in the document]
Comment (gudbrandtandberg) [reply src/tlc_tools/datasets.py:72]: Any tables currently created by 3lc will have the "bbs.bb_list.label" layout, but this is of course open for future changes.

File: src/tlc_tools/metrics.py [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+    final_traversal_order = [unique_indices[idx] for idx in traversal_order]
+
+    # Add duplicates in random order
+    duplicate_indices = [i for i in range(len(embeddings)) if i not in unique_indices]
+    np.random.shuffle(duplicate_indices)
+    final_traversal_order.extend(duplicate_indices)
+
+    return [int(idx) for idx in final_traversal_order]
+
+IMAGE_METRICS = Literal["width", "height", "brightness", "contrast", "sharpness", "average_red", "average_green", "average_blue"]
```
[the remark this answers is not in the document]
Comment (gudbrandtandberg) [reply src/tlc_tools/metrics.py]: This was not so straight forward, as Literal[ALL_IMAGE_METRICS] will not be valid since it is an expression in the literal.. If I have time I will try to find a way to not repeat myself.

File: src/tlc_tools/metrics.py [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+    
+    if "width" in metrics:
+        computed_metrics["width"] = width
+
+    if "height" in metrics:
+        computed_metrics["height"] = height
+
+    pixels = np.array(image)
+
+    # Convert to grayscale for some metrics
```
[the remark this answers is not in the document]
Comment (gudbrandtandberg) [reply src/tlc_tools/metrics.py:178]: Nice, did it.

File: src/tlc_tools/weights.py [hunk 15]
```diff
@@ -0,0 +1,39 @@
+import numpy as np
+from collections import defaultdict
+
+import tlc
+
+
+def object_detection_sample_weights(table: tlc.Table, alpha=1.0, include_zero_weighted_images=False):
```
[the remark this answers is not in the document]
Comment (gudbrandtandberg) [reply src/tlc_tools/weights.py]: Ok, decided to remove and stash ;)

File: tests/unit/test_add_columns_to_table.py [hunk 16]
```diff
@@ -0,0 +1,2 @@
+def test_add_columns_to_table():
+    assert 1 == 1
```
[the remark this answers is not in the document]
Comment (gudbrandtandberg) [reply tests/unit/test_add_columns_to_table.py]: Removed

File: tutorials/1-create-tables/README.md
[the remark this answers is not in the document]
Comment (gudbrandtandberg) [reply tutorials/1-create-tables/README.md:1]: Yes, we strive to use the same capitalization here as in monorepo. Have fixed everything I came across.

File: src/tlc_tools/datasets.py [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+            )
+            proposal_box = [x, y, w, h]
+
+            # Ensure the proposed box does not intersect any ground truth boxes
+            if not any(self._intersects(proposal_box, gt_box) for gt_box in gt_boxes_xywh):
+                break
+
+        if i == max_attempts - 1:
+            # Return a 100x100 black square if no valid background patch is found
+            return Image.new("RGB", (100, 100), color=(0, 0, 0)), torch.tensor(self.background_label, dtype=torch.long)
```
[the remark this answers is not in the document]
Comment (gudbrandtandberg) [reply src/tlc_tools/datasets.py:169]: I feel OK leaving this up to the user to change if it is a problem (added a warning). 
It certainly can happen, but it is quite degenerate and if it happens frequently, the user should really not be sending include_background=True (because their images are in this case filled with bounding boxes, so finding backgrounds is not a good idea.

File: src/tlc_tools/split.py [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+
+        # FPS one split at a time
+        split_indices = {}
+        for split_name, split_proportion in list(splits.items())[:-1]:
+            # Determine the number of samples for this split
+            split_size = int(len(original_indices) * split_proportion)
+
+            # Perform sampling on the current subset of remaining indices
+            sampled_indices = fpsample.bucket_fps_kdtree_sampling(
+                by_column[remaining_indices],  # Subset of `by_column` corresponding to remaining indices
```
[the remark this answers is not in the document]
Comment (gudbrandtandberg) [reply src/tlc_tools/split.py:89]: Fred added a raise if by_column is none

File: tutorials/1-create-tables/README.md [hunk 19]
```diff
@@ -6,13 +6,13 @@ The `tlc` Python package provides several helper functions for creating tables f
...
 | Create a table from a folder of subfolders, each subfolder containing images belonging to a certain class. | Write a custom Table by specifying a schema and adding data row-by-row. | Write a custom Table containing images and bounding boxes, using custom bounding boxes of an arbitrary format. |
-| **YOLO** | **Semantic Segmentation**  | **Video Thumbnails** |
+| **YOLO** | **Semantic segmentation**  | **Video thumbnails** |
 | [![yolo][yolo-img]][yolo-link] | [![semseg][semseg-img]][semseg-link] | [![video][video-img]][video-link] |
 | Load images and bounding boxes from the YOLO YAML file format. | Write a Table from images and one or more masks. | Work with video datasets in 3LC by creating thumbnails and storing URLs to video files. |
-| **From Torch Dataset** | **COCO** |    |
+| **From PyTorch Dataset** | **COCO** |    |
 | [![torch][torch-img]][torch-link] | [![coco][coco-img]][coco-link] |    |
 | Create a Table directly from a PyTorch Dataset (CIFAR-10). | Create a table from a folder of images and a COCO format JSON file. | |
```
Comment (KellyWalker) [inline tutorials/1-create-tables/README.md:17]:
```suggestion
| Create a Table directly from a PyTorch Dataset (CIFAR-10). | Create a Table from a folder of images and a COCO format JSON file. | |
```

File: src/tlc_tools/metrics.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (KellyWalker) [reply src/tlc_tools/metrics.py:178]: Maybe I am missing something, but this code does not seem to have been moved

File: src/tlc_tools/embeddings.py [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+    :param model: Model to use for embedding extraction. If None, defaults to ViT from Hugging Face.
+    :param embedding_extraction_fn: Function to extract embeddings from the model output. If None, defaults to selecting the [CLS] token for ViT.
+    :param embedding_column: Column name to add embeddings to in the table.
+    :param batch_size: Batch size for processing images.
+    :param device: Device to use for inference.
+    :param preprocess_fn: Preprocessing function for images. If None, defaults to a standard ViT preprocessing pipeline.
```
[the remark this answers is not in the document]
Comment (KellyWalker) [reply src/tlc_tools/embeddings.py:36]: I do not see the docstring for `preprocess_fn` reflecting what I wrote.

File: src/tlc_tools/metrics.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (gudbrandtandberg) [reply src/tlc_tools/metrics.py:167]: I had changed it to only compute pixel values if necessary, and only compute the RGB values if necessary.. Should be better, can you check? 

Also managed to get a list of the Literals values using get_args, so no longer repeating those metric names :)

File: src/tlc_tools/embeddings.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (gudbrandtandberg) [reply src/tlc_tools/embeddings.py]: Very strange, it was unstaged on my side, unclear how that was possible.. Perhaps got rejected by pre-commit and I didnt notice or something.

**Review by fredmell — APPROVED** [review]

File: src/tlc_tools/metrics.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (fredmell) [reply src/tlc_tools/metrics.py:167]: Gudbrand will triple check tomorrow. Resolving for now.
